### PR TITLE
[DRAFT] [SNOW-1625268] Add github action to ensure AST correctness

### PR DESCRIPTION
**Pull request by sfc-gh-lspiegelberg**
tbd,

### CODE REVIEW

File: tests/integ/test_dataframe.py [hunk 1]
```diff
@@ -175,7 +175,7 @@ def test_read_stage_file_show(session, resources_path, local_testing_mode):
             session.read.option("purge", False)
             .schema(user_schema)
             .csv(test_file_on_stage)
```
Comment (sfc-gh-lspiegelberg) [inline tests/integ/test_dataframe.py:177]: `_show_string` is an internal API. Therefore pass explicitly `_emit_ast`. We do not mark internal APIs with `@publicapi`.

File: src/snowflake/snowpark/dataframe_na_functions.py [hunk 2]
```diff
@@ -392,7 +392,8 @@ def fill(
         if not value_dict:
             new_df = copy.copy(self._dataframe)
             add_api_call(new_df, "DataFrameNaFunctions.fill")
-            new_df._ast_id = stmt.var_id.bitfield1
+            if _emit_ast:
+                new_df._ast_id = stmt.var_id.bitfield1
```
Comment (sfc-gh-oplaton) [inline src/snowflake/snowpark/dataframe_na_functions.py:396]: Intuitively, this pattern seems suspicious. Is it OK for two distinct object to share the same `VarId`?

File: tests/integ/scala/test_dataframe_join_suite.py
Comment (sfc-gh-oplaton) [inline tests/integ/scala/test_dataframe_join_suite.py:1]: This, and other test files that follow the same pattern are surprising: Why don't `_show_string` and other methods involved take the default value of `_emit_ast` from the session, the same way as public APIs? What prevents that from working? (Or why is it a bad idea, if it is?)

Comment (sfc-gh-lspiegelberg) [reply tests/integ/scala/test_dataframe_join_suite.py:1]: That's a great question! My reasoning was here as follows:

1.) `_show_string` is a private API. It is not exposed to the user, and therefore doesn't get a `@publicapi` decorator.
2.) Because it is private, the caller of this API is responsible for the state (i.e., passing `_emit_ast`).
3.) Unfortunately, in the tests the API design is not correctly accounted for. A private API is used to test things. The contract of `show()` is to output to stdout. Hence, a test should capture stdout and compare. However, for whichever reason this is not done. Instead a private (!) function is used to write end-to-end integration tests. I saw 3 options:
a.) Make `_show_string()` a public API, that doesn't great.
b.) Rewrite all tests to use `show()` and stdout capture -> a lot of work.
c.) Go a hybrid, basically honor the original test intention but have it do the private contract: Pass `_emit_ast` and ensure state because internals are getting called. -> that's what I did.

to sum up, the correct way to do it for an integration test would be:
1.) make `_show_string()` a public API OR
2.) do not test internal functions in integration tests. Internals are tested with unit tests at most, never integration tests. Else, you screwed up API design.

File: tests/integ/test_dataframe.py (hunk 1 above)
Comment (sfc-gh-lspiegelberg) [reply tests/integ/test_dataframe.py:178]: @sfc-gh-oplaton I tried to explain the reasoniing here.

**Review by sfc-gh-oplaton — APPROVED** [review]

File: src/snowflake/snowpark/dataframe_na_functions.py (hunk 2 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/dataframe_na_functions.py:396]: that's a good question, here I basically simply assigned (like in other APIs) the id of the stmt to the new dataframe returned. The if stmt is needed, as the code is mixed (in some APIs we have a clear separation of AST code vs. existing, in others it is mixed in because of some parameter processing).
